Runtime option for tuning

DavidPoliakoff opened the pull request:
Implements #3454 .

The tuning functionality is highly experimental. Further, people could want to do tuning, but not want Kokkos to tune itself (or user policies). This PR separates the idea of "the Kokkos Tuning infrastructure exists" from "Kokkos tunes itself," the latter of which is a runtime option. I introduced it as a default-off option, though I'm flexible on that

DavidPoliakoff commented:
Retest this please

dalg24 commented on core/src/impl/Kokkos_Core.cpp:
```diff
@@ -273,6 +274,7 @@ void initialize_profiling(const InitArguments&) {
 
 void pre_initialize_internal(const InitArguments& args) {
   if (args.disable_warnings) g_show_warnings = false;
+  if (args.tune_kokkos_internals) g_tune_internals = false;
```
I'd expect that would set to `true`

DavidPoliakoff replied:
Oooooooof. Good catch

dalg24 commented on core/src/Kokkos_Core.hpp:
```diff
@@ -81,14 +81,16 @@ struct InitArguments {
   bool disable_warnings;
-
-  InitArguments(int nt = -1, int nn = -1, int dv = -1, bool dw = false)
+  bool tune_kokkos_internals;
```
Why did you specify "kokkos" in the name?

DavidPoliakoff replied:
It's just how I think of it, this is describing whether Kokkos tunes its internals. I don't feel too strongly about it

dalg24 replied:
I commented because it feels inconsistent to me next to `g_tune_internals` and to `tune_internals()`

DavidPoliakoff replied:
Works for me, pushed 9c8023d

dalg24 commented on core/src/impl/Kokkos_Core.cpp:
```diff
@@ -512,6 +522,7 @@ void parse_command_line_arguments(int& narg, char* arg[],
 
       --kokkos-help                  : print this message
       --kokkos-disable-warnings      : disable kokkos warning messages
+      --kokkos-tune-internals        : allow Kokkos to do autotuning
```
I keep starring at it and I am still not sure I actually understand what the option does.

dalg24 commented on core/src/impl/Kokkos_Profiling.hpp:
```diff
@@ -54,6 +54,9 @@
 #include <map>
 #include <type_traits>
 namespace Kokkos {
+// forward declaration
+
+bool tune_internals() noexcept;
```
```suggestion

// forward declaration
bool tune_internals() noexcept;

```

dalg24 commented on core/src/impl/Kokkos_Core.cpp:
```diff
@@ -932,6 +957,7 @@ void print_configuration(std::ostream& out, const bool detail) {
 bool is_initialized() noexcept { return g_is_initialized; }
 
 bool show_warnings() noexcept { return g_show_warnings; }
+bool tune_internals() noexcept { return g_tune_internals; }
```
Did you mean to add in `Kokkos::`?

DavidPoliakoff replied:
I discussed this yesterday. I want `show_warnings` and `tune_internals` to work the same way, they both query something about Kokkos. I'm comfortable with either of two options

1) Leave both in Kokkos
2) Put tune_internals in Kokkos::Impl, same for show_warnings (deprecating the show_warnings in namespace Kokkos)

dalg24 replied:
I would prefer `Impl::` and deprecate/move `show_warnings` in a separate PR.

crtrott replied:
I think it makes some sense to have them in Kokkos:: i.e. to let users check whether this is on or not.

DavidPoliakoff replied:
I have no strong feelings one way or the other. You two have a discussion and see how you feel?

DavidPoliakoff replied:
I thought a bit more about it, and I'd like to leave the function in namespace Kokkos (no impl). Some things a user might do

"If the user didn't tell Kokkos to tune, I won't tune either" 
"if the user told Kokkos to tune, I don't need to tune such-and-such a parameter because it's analogous to one in Kokkos"

dalg24 commented on core/src/impl/Kokkos_Core.cpp:
```diff
@@ -512,6 +522,9 @@ void parse_command_line_arguments(int& narg, char* arg[],
       --kokkos-disable-warnings      : disable kokkos warning messages
+      --kokkos-tune-internals        : allow Kokkos to autotune policies and declare
+                                       tuning features through the tuning system. If
+				       left off, Kokkos uses heuristics
```
```suggestion
				         left off, Kokkos uses heuristics
```

dalg24 reviewed (commented):
I am pretty much good with it modulo history clean up

crtrott reviewed (changes requested):
Looks good. But can we add a simple test to the initialization tests? I.e. something which sets the InitArgument and then checks that tune_internals has the right value?

DavidPoliakoff commented:
@crtrott : addressed in 4614a23, but do review the test changes. Our free'ing strategy leaked memory when we introduced arguments which Kokkos shifted (as it does with kokkos-tune-internals), I fixed that

crtrott commented on core/src/impl/Kokkos_Core.cpp:
```diff
@@ -704,6 +717,20 @@ void parse_environment_variables(InitArguments& arguments) {
+    }
+    if ((env_str == "TRUE") || (env_str == "ON") || (env_str == "1"))
+      tune_internals = true;
+    else if (!tune_internals)
```
I think this logic is wrong. You are entering the environment variable is set to false branch, which means if !tune_internals everything is dandy. Note that even if you fix this (else if (tune_internals) without the !) there is still an issue which is also present for disable warnigns (thinking through the logic made that clear to me): if the user explicitly sets disable_warnings in the command line arguments to false, and also sets the enviornment variable to true, it will accept the environment variable, because there is no way to distinguish between "not set" and "set to false" for the command line arguments. We probably want to fix that some time. But for now I would be ok to just remove the ! here and then this should be good.

DavidPoliakoff replied:
Yeah, you're right (also, somehow I missed this). Really good catch

DavidPoliakoff replied:
Addressed in 1cc5bb6

crtrott reviewed (changes requested):
See comment on check logic.

DavidPoliakoff commented:
@dalg24, re history cleanup, comfortable just squash-merging? No other work depends on this
